linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+            AB[i][j] = AB[i][j] + A[i][k] * B[k][j];
+         }
+      }
+   }
+   return AB;
+}
+
+/**
+ * @overload
+ * @note vector . matrix
+ */
+template <typename S, typename T, int m, int n> MFEM_HOST_DEVICE
+auto dot(const tensor<S, m>& A, const tensor<T, m, n>& B) ->
+tensor<decltype(S {} * T{}), n>
+{
+   tensor<decltype(S{} * T{}), n> AB{};
+   for (int i = 0; i < n; i++)
+   {
+      for (int j = 0; j < m; j++)
+      {
+         AB[i] = AB[i] + A[j] * B[j][i];
+      }
+   }
+   return AB;
+}
+
+/**
+ * @overload
+ * @note matrix . vector
+ */
+template <typename S, typename T, int m, int n> MFEM_HOST_DEVICE
+auto dot(const tensor<S, m, n>& A, const tensor<T, n>& B) ->
+tensor<decltype(S {} * T{}), m>
+{
+   tensor<decltype(S{} * T{}), m> AB{};
+   for (int i = 0; i < m; i++)
+   {
+      for (int j = 0; j < n; j++)
+      {
+         AB[i] = AB[i] + A[i][j] * B[j];
+      }
+   }
+   return AB;
+}
+
+/**
+ * @overload
+ * @note 3rd-order-tensor . vector
+ */
+template <typename S, typename T, int m, int n, int p> MFEM_HOST_DEVICE
+auto dot(const tensor<S, m, n, p>& A, const tensor<T, p>& B) ->
+tensor<decltype(S {} * T{}), m, n>
+{
+   tensor<decltype(S{} * T{}), m, n> AB{};
+   for (int i = 0; i < m; i++)
+   {
+      for (int j = 0; j < n; j++)
+      {
+         for (int k = 0; k < p; k++)
+         {
+            AB[i][j] += A[i][j][k] * B[k];
+         }
+      }
+   }
+   return AB;
+}
+
+// /**
+//  * @brief Dot product of a vector . vector and vector . tensor
+//  *
+//  * @tparam S the underlying type of the tensor (lefthand) argument
+//  * @tparam T the underlying type of the tensor (righthand) argument
+//  * @tparam m the dimension of the first tensor
+//  * @tparam n the parameter pack of dimensions of the second tensor
+//  * @param A The lefthand tensor
+//  * @param B The righthand tensor
+//  * @return The computed dot product
+//  */
+// template <typename S, typename T, int m, int... n>
+// auto dot(const tensor<S, m>& A, const tensor<T, m, n...>& B)
+// {
+//    // this dot product function includes the vector * vector implementation and
+//    // the vector * tensor one, since clang emits an error about ambiguous
+//    // overloads if they are separate functions. The `if constexpr` expression avoids
+//    // using an `else` because that confuses nvcc (11.2) into thinking there's not
+//    // a return statement
+//    if constexpr (sizeof...(n) == 0)
+//    {
+//       decltype(S{} * T{}) AB{};
+//       for (int i = 0; i < m; i++)
+//       {
+//          AB += A[i] * B[i];
+//       }
+//       return AB;
+//    }
+
+//    if constexpr (sizeof...(n) > 0)
+//    {
+//       constexpr int                     dimensions[] = {n...};
+//       tensor<decltype(S{} * T{}), n...> AB{};
+//       for (int i = 0; i < dimensions[0]; i++)
+//       {
+//          for (int j = 0; j < m; j++)
+//          {
+//             AB[i] = AB[i] + A[j] * B[j][i];
+//          }
+//       }
+//       return AB;
+//    }
+// }
```

Comment:
Why is this commented out? Is it an alternative implementation for what comes next?

Reply:
Yes exactly. As soon as mfem moves to a reasonable cpp standard, we can delete a huge part of specialized implementations.

Reply:
Makes sense. Can you put a comment at the top of this commented-out block saying that?

Reply:
The last comment is not addressed yet.